File: leetcode/131. Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def is_palindome(start, end):
            while start < end:
                if s[start] != s[end]:
                    return False
                start += 1
                end -= 1
            return True
        
        res = []
        def helper(idx, curr_lst):
            if idx == len(s):
                res.append(curr_lst[:])
                return
            
            for i in range(idx, len(s)):
                if is_palindome(idx, i):
                    curr_lst.append(s[idx:i+1])
                    helper(i+1, curr_lst)
                    curr_lst.pop()
        helper(0, [])
        return res

File: leetcode/131. Palindrome Partitioning.kt
class Solution {
    fun partition(s: String): List<List<String>> {
        fun isPalindrome(i: Int, j: Int): Boolean {
            var i = i
            var j = j
            while (i < j) {
                if (s[i] != s[j]) return false
                i += 1
                j -= 1
            }
            return true
        }
        val res = mutableListOf<List<String>>()
        fun get_partitions(start: Int, curr: MutableList<String>) {
            if (start == s.length) {
                res.add(curr.toList())
                return
            }
            for (i in start..s.length-1) {
                if (!isPalindrome(start, i)) continue
                val sub = s.substring(start, i+1)
                curr.add(sub)
                get_partitions(i+1, curr)
                curr.removeLast()
            }
        }
        get_partitions(0, mutableListOf<String>())
        return res
    }
}
